### PR TITLE
Attribute Cache Map+Tree

### PR DESCRIPTION
**Pull request by foodprocessor**
### What type of Pull Request is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

### Describe your changes in brief
Use both a map and a tree to manage attribute cache items

### Checklist

- [ ] Tested locally
- [ ] Added new dependencies
- [ ] Updated documentation
- [ ] Added tests

### Related Issues

- Related Issue #
- Closes #

### CODE REVIEW

**Comment by foodprocessor** [comment]
Should I rename cacheMap.go?
Maybe to treeMap.go?

File: component/attr_cache/attr_cache.go [hunk 1]
```diff
@@ -164,50 +163,56 @@ func (ac *AttrCache) OnConfigChange() {
...
-
-	// Delete all descendants of the path, then delete the path
-
-	//get attrCacheItem
-	toBeDeleted, getErr := ac.cacheMap.get(path)
-	if getErr != nil || !toBeDeleted.exists() {
-		log.Err("AttrCache::deleteCachedDirectory : %s not found", path)
+	// get the item
+	item, found := ac.cache.get(path)
+	if !found || !item.exists() {
```
Comment (Dabnsky) [inline component/attr_cache/attr_cache.go:194]: I see you are using the ```||``` comparison here but you have previously split this up into an ```else if``` in deleteDirectory(). Is this trivial? is it for more readability? I'm just noticing the style of error checking is not too consistent so far

File: component/attr_cache/attr_cache.go [hunk 2]
```diff
@@ -164,50 +163,56 @@ func (ac *AttrCache) OnConfigChange() {
...
-	toBeDeleted, getErr := ac.cacheMap.get(path)
-	if getErr != nil || !toBeDeleted.exists() {
+
+// Record that the directory and all its contents have been deleted from the file system.
+// This should only be called when cacheDirs is false.
+// If cacheDirs is true, use deleteCachedDirectory instead.
+func (ac *AttrCache) deleteDirectory(path string, deletedAt time.Time) {
+	// get the entry to be marked deleted
+	value, found := ac.cache.get(path)
+	if !found {
```
Comment (Dabnsky) [inline component/attr_cache/attr_cache.go:173]: just calling it ```found``` is very elegant

File: component/attr_cache/attr_cache.go (hunk 1 above)
Comment (foodprocessor) [reply component/attr_cache/attr_cache.go:194]: In `deleteDirectory`, I chose to take two different actions for the two cases:
- if there is no entry at all (`!found`), I add a new entry which is marked deleted
- if there is an entry which is marked deleted (`!value.exists()`), I do nothing (but log a warning)
But here I do the same thing for both cases (log and return an error).

File: component/attr_cache/attr_cache.go (hunk 2 above)
Comment (foodprocessor) [reply component/attr_cache/attr_cache.go:173]: Thank you!
This was inspired by Go's map:
`value, ok := mapVar["key"]`

**Review by jfantinhardesty — APPROVED** [review]

File: component/attr_cache/cacheMap.go [hunk 3]
```diff
@@ -135,7 +144,7 @@ func (value *attrCacheItem) isInCloud() bool {
 }
 
 func (value *attrCacheItem) markDeleted(deletedTime time.Time) {
-	if !value.isDeleted() {
+	if value.exists() {
```
[the remark this answers is not in the document]
Comment (foodprocessor) [reply component/attr_cache/cacheMap.go:147]: These functions both pointed to the same flag (`AttrFlagExists`). I removed the function `isDeleted()`.

File: component/attr_cache/attr_cache.go [hunk 4]
```diff
@@ -164,50 +163,56 @@ func (ac *AttrCache) OnConfigChange() {
...
-// That way if a request came in for a deleted item, we can respond from the cache.
-func (ac *AttrCache) deleteDirectory(path string, time time.Time) {
-	//get attrCacheItem
-	toBeDeleted, getErr := ac.cacheMap.get(path)
-	if getErr != nil || !toBeDeleted.exists() {
+
+// Record that the directory and all its contents have been deleted from the file system.
+// This should only be called when cacheDirs is false.
+// If cacheDirs is true, use deleteCachedDirectory instead.
+func (ac *AttrCache) deleteDirectory(path string, deletedAt time.Time) {
```
[the remark this answers is not in the document]
Comment (foodprocessor) [reply component/attr_cache/attr_cache.go:170]: Good point! I'll try and see how it goes.

[the remark this answers is not in the document]
Comment (foodprocessor) [reply component/attr_cache/attr_cache.go]: Done! I'm glad you suggested this!

**Review by Ferelith-maker — APPROVED** [review]

**Pull request merged**